552 add access to item poses in mujoco scenes

ClaireHzl opened the pull request:
It's more of a proposal for discussion; the architecture is open to debate in terms of how we want to implement it in the Client SDK.

**Goes with branches:**
- [237 of reachy2_sdk_server](https://github.com/pollen-robotics/reachy2_sdk_server/tree/237-add-access-to-item-poses-in-mujoco-scenes) 
- [188 of reachy2-sdk-api](https://github.com/pollen-robotics/reachy2-sdk-api/tree/188-add-access-to-item-poses-in-mujoco-scenes)
- [4 of reachy2_mujoco_assets](https://github.com/pollen-robotics/reachy2_mujoco_assets/tree/4-add-tracked-objects)

**Mujoco scenes side**:
Definition of the desired tracked bodies. 

**Server side:** 
- Addition of a subscriber to the `mujoco/item_position` topic, which updates a dictionary of poses for objects that are tracked continuously. 
- Addition of a Mujoco servicer, active only if the core_mode is Mujoco

**Client side:** 
- Added a Mujoco module if core_mode is Mujoco
- available via `reachy.mujoco`
- Three functions: 
  - one to obtain the entire dictionary of tracked object poses
  - one to obtain a specific object
  - one to obtain the relative pose of an object in Reachy's frame
  
  To test it, you can launch with fruits scene : 
 ```bash
  reset; ros2 launch reachy_bringup reachy.launch.py start_sdk_server:=true start_rviz:=true fake:=true mujoco:=true scene:=fruits
```
then check `reachy.mujoco.get_objects_poses()`, `reachy.mujoco.get_object_pose('orange1')` and `reachy.mujoco.get_relative_object_pose('orange1')`

ClaireHzl committed:
Add Mujoco module.

ClaireHzl committed:
Add functions.

ClaireHzl committed:
Fix mypy errors.

ClaireHzl committed:
Fix pydoc issue.

glannuzel commented:
We will need to change minimum required version for reachy2-sdk-api here: https://github.com/pollen-robotics/reachy2-sdk/blob/2b91b165c10eb64dd4947813bd5035b1135ca45e/setup.cfg#L24

glannuzel commented:
And update the sdk version: https://github.com/pollen-robotics/reachy2-sdk/blob/2b91b165c10eb64dd4947813bd5035b1135ca45e/src/reachy2_sdk/__init__.py#L31

glannuzel reviewed (commented):
The calls to the 3 functions work well!

I think I would prefer having to import something else rather than use directly reachy2_sdk and have reachy.mujoco available.

I am not sure I did use correctly the relative_pose, I thought I would be able to use those poses with the arms goto functions, but the result was not as expected. I will retry this!

Can we have a function that returns the list of the tracked objects as well? :smile:

ClaireHzl committed:
Add functions.

ClaireHzl committed:
Separate reachy2sdk and reachy2mujoco.

ClaireHzl commented:
I changed the architecture, reachy2_mujoco is separated from reachy2_sdk
to import it : `from reachy2_mujoco import ReachyMujoco`
to instanciate it : `r_mujoco = ReachyMujoco(reachy)`

I added the function get_tracked_objects_list() and reworked the function to get the relative pose.
I added a function in utils that takes a position and a dictionary of intrinsic rotations to create a pose (I find it useful, but let's see what you think)
I changed the version of the api to 1.0.21 and the sdk to 1.0.15

glannuzel commented on src/reachy2_sdk/utils/utils.py:
```diff
@@ -602,3 +602,53 @@ def get_normal_vector(vector: npt.NDArray[np.float64], arc_direction: str) -> Op
     return normal.astype(np.float64)
+
+
+def create_goal_pose_from_rotations(
```
I think this function does the same as `get_pose_matrix()` above in the file 
We probably already have the necessary utils to do create pose to grasp objects in mujoco thanks to `decompose_matrix()` and `get_pose_matrix()`

glannuzel commented on src/reachy2_mujoco/reachy_mujoco.py:
```diff
@@ -0,0 +1,118 @@
+            grpc_channel: The gRPC channel to communicate with the MuJoCo service.
+        """
+        if reachy._mode != "MUJOCO":
+            raise ValueError(f"You are in {reachy._mode} mode, not MUJOCO.")
```
Maybe say "Reachy is in..." rather than "You are..."?

glannuzel reviewed (approved):
Seems to work very well!
I managed to use it to reach oranges and apples in the fruit scene.

I am just wondering if the name "reachy2_mujoco" is not going to be confusing (makes think it replaces reachy2_sdk in mujoco mode). Should we be more specific, like "reachy2_mujoco_scene"?

ClaireHzl committed:
Change name of mujoco descriptor.

ClaireHzl commented:
I change the Mujoco module name for MujocoSceneDescription. 

And I decided to add the utils function for the goal pose in the next version of SDK, with the rework of goto tolerance (to make the changes more relevant and clear).
